core/src/apps/zcash/orchard/crypto/note_encryption.py

```diff
@@ -0,0 +1,120 @@
+import gc
+from micropython import const
+from typing import TYPE_CHECKING
+
+from trezor.crypto import chacha20poly1305
+from trezor.crypto.hashlib import blake2b
+from trezor.crypto.pallas import Point
+from trezor.utils import empty_bytearray, ensure
+
+from apps.common.writers import write_bytes_fixed
+
+from .note import Note
+
+if TYPE_CHECKING:
+    from typing import Iterable
+    from ..random import ActionShieldingRng
+
+BLOCK_SIZE = const(64)
+ENC_CIPHERTEXT_SIZE = const(580)
+OUT_CIPHERTEXT_SIZE = const(80)
+
+
+# https://zips.z.cash/protocol/nu5.pdf#concreteorchardkdf
+def kdf_orchard(shared_secret: Point, ephemeral_key: bytes) -> bytes:
+    digest = blake2b(outlen=32, personal=b"Zcash_OrchardKDF")
+    digest.update(shared_secret.to_bytes())
+    digest.update(ephemeral_key)
+    return digest.digest()
+
+
+# https://zips.z.cash/protocol/nu5.pdf#concreteprfs
+def prf_ock_orchard(
+    ovk: bytes,
+    cv: bytes,
+    cmx: bytes,
+    ephemeral_key: bytes,
+) -> bytes:
+    digest = blake2b(outlen=32, personal=b"Zcash_Orchardock")
+    digest.update(ovk)
+    digest.update(cv)
+    digest.update(cmx)
+    digest.update(ephemeral_key)
+    return digest.digest()
+
+
+def chunks(size: int, length: int) -> Iterable[tuple[int, int]]:
+    offset = 0
+    while offset + size <= length:
+        yield offset, offset + size
+        offset += size
+    if offset < length:
+        yield offset, length
+    return
```

Comment:
to avoid making copies of every chunk of the buffer:
```suggestion
def chunks(size: int, buffer: bytes) -> Iterable[memoryview]:
    offset = 0
    m = memoryview(buffer)
    while offset + size <= length:
        yield m[offset : offset + size]
        offset += size
    if offset < length:
        yield m[offset : length]
```

Reply:
commited